infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvironmentVariablesSorter.java

```diff
@@ -0,0 +1,202 @@
+
+    // we don't need to check for cycle here, because that has been detected when establishing
+    // the closure.
+    if (aTransRefs.contains(b.getName())) {
+      // a depends on b, hence a > b.
+      return 1;
+    } else if (bTransRefs.contains(a.getName())) {
+      // b depends on a -> a < b.
+      return -1;
+    }
+
+    // ok, the two env vars don't depend on each other, so we actually don't care too much about
+    // the order. We need to be consistent and total, though...
+
+    int sizeDiff = aTransRefs.size() - bTransRefs.size();
+    if (sizeDiff != 0) {
+      return sizeDiff;
+    }
+
+    // ah, the two env vars have the same number of references, so we need to take a "slow path"
+    // in determining the total order.
+    Iterator<String> aIt = aRefs.iterator();
+    Iterator<String> bIt = bRefs.iterator();
+    while (aIt.hasNext()) {
+      if (!bIt.hasNext()) {
+        // a has more references than b
+        return 1;
+      }
+
+      String aRef = aIt.next();
+      String bRef = bIt.next();
+
+      int res = aRef.compareTo(bRef);
+      if (res != 0) {
+        return res;
+      }
+    }
+
+    return 0;
+  }
+
+  private SortedSet<String> transitiveClosure(String variableName) {
+    SortedSet<String> recRefs = recursiveReferenceMap.get(variableName);
+    if (recRefs == null) {
+      recRefs = new TreeSet<>(naturalOrder());
+      fillRecursiveReferences(variableName, referenceMap, recRefs);
+      recursiveReferenceMap.put(variableName, recRefs);
+    }
+    return recRefs;
+  }
+
+  @VisibleForTesting
+  static SortedSet<String> findReferences(EnvVar var) throws CycleException {
+    String val = var.getValue();
+
+    Matcher matcher = REFERENCE_PATTERN.matcher(val);
+
+    SortedSet<String> ret = new TreeSet<>(naturalOrder());
+
+    while (matcher.find()) {
+      int start = matcher.start();
+
+      // the variable reference can be escaped using a double $, e.g. $$(VAR) is not a reference
+      if (start > 0 && val.charAt(start - 1) == '$') {
+        continue;
+      }
+
+      // extract the variable name out of the reference $(NAME) -> NAME
+      String refName = matcher.group().substring(2, matcher.group().length() - 1);
+      if (refName.equals(var.getName())) {
+        throw new CycleException(refName);
+      }
+      ret.add(refName);
+    }
+
+    return ret;
+  }
+
+  private static void fillRecursiveReferences(
+      String name, Map<String, SortedSet<String>> directReferences, SortedSet<String> result)
+      throws CycleException {
+
+    Set<String> nameRefs = directReferences.get(name);
+    if (nameRefs == null || nameRefs.isEmpty()) {
+      return;
+    }
+
+    if (result.contains(name)) {
+      throw new CycleException(name);
+    }
+
+    for (String refName : nameRefs) {
+      result.add(refName);
+      fillRecursiveReferences(refName, directReferences, result);
+    }
+  }
+
+  static final class CycleException extends RuntimeException {
```

Comment:
This is thrown from `compare` which must not throw a checked exception by contract.

Reply:
I don't think it should be thrown from the compare method. Would rather refactor code to do dependencies and cycles processing in one method and call sort with the comparator in another

Reply:
I don't even mean public methods, just restructure the code to respect single responsibility principle

Reply:
I do the cycle detection lazily on purpose actually, because theoretically, it can be a lot of work. So I do it only once and only if necessary.

Depending on how the sorting algorithm is implemented it can find the offending cycle much sooner than we could if we went through the whole collection looking for them. This is actually just a bet that the more complex "traversal" of the list that the sorting algorithm is performing is going to bump into the possible offenders that happen to be located near the end of the input collection sooner than a linear traversal would.
 
As for the single responsibility principle, I guess I still don't see how this class breaks it. Could you please explain that in more detail?

Reply:
It provides util method to sort env vars (1 responsibility) and it implements a comparing of items (2 responsibility)

Reply:
As far as I understand there should be done quite the same work to find cyclic dependencies and sort env var.
I think `CycleException` solution is good enough.